lib/upload.js

```diff
@@ -1014,6 +1097,96 @@ function splitSizeIntoParts (totalSize, partCount) {
   return parts
 }
 
+/**
+ * Build the chunk staggers.
+ *
+ * NOTE: Must be kept in sync with `splitSizeIntoParts` in skynet-js.
+ */
+function buildChunkStaggers (totalSize, partCount, chunkSize) {
+  // The lists of chunks for each part.
+  const partsChunks = []
+  for (let i = 0; i < partCount; i++) {
+    partsChunks.push([])
+  }
+  // The leftover size that must go into the last part.
+  const hasLeftover = (totalSize % chunkSize) > 0
+
+  function newChunk (index) {
+    const partChunks = partsChunks[index]
+
+    let chunkResolve
+    const promise = new Promise((resolve) => chunkResolve = resolve)
+    const chunk = {
+      promise,
+      resolve         : chunkResolve,
+      resolved        : false,
+      nextChunkStagger: null,
+      nextChunkInPart : null,
+      indexInPart     : partChunks.length,
+    }
+
+    // Previous chunk in part should point to this chunk.
+    if (partChunks.length > 0) {
+      partChunks[partChunks.length - 1].nextChunkInPart = chunk
+    }
+
+    // Push chunk to part.
+    partChunks.push(chunk)
+  }
+
+  // Assign chunks to parts in order, looping back to the beginning if we get to
+  // the end of the parts array.
+  let lastPart = -1
+  for (let i = 0; i < Math.floor(totalSize / chunkSize); i++) {
+    newChunk(i % partCount)
+
+    if (i > lastPart) lastPart = i
```

Comment:
you are not breaking from the `for` loop here right, doesn't that mean you know the `lastpart` beforehand?